### ISSUE TITLE
How do I retain my language settings while upgrading?

### ISSUE THREAD

**Issue by iainelder** [dbeaver/dbeaver#6872]
#### System information: 
- Ubuntu 18.04
- Dbeaver Version 6.2.1.201909221630

#### Describe the problem you're observing:

My desktop user interface language is Spanish.

I want to use DBeaver in English.

I had set DBeaver's UI language to English by editing `/usr/share/dbeaver/dbeaver.ini`as instructed in the [Localization wiki](https://github.com/dbeaver/dbeaver/wiki/Localization).

I upgraded to muy current version of DBeaver by following the prompt at startup and installing a new .deb package.

After the upgrade the UI langauge reverted to Spanish.

I had to edit the ini file again to restore my settings.

How do I stop DBeaver from resetting the language when it upgrades?

**Comment by ajperez81** [comment]
I have the same problem.

My OS is setup in Spanish but I prefer using DBeaver in English. Whenever I update the software (manually, using the zipped version), the language is reset back to Spanish. I have to change it to English again (in the preferences dialog, I don't edit the .ini file directly) and then restart DBeaver for the changes to be applied. All my other settings, saved connections, etc., are properly carried from one version to the other, the only thing is lost is the language.

It's annoying, specially because DBeaver is updated so frequently (almost twice a month... amazing job improving the software! :+1:)

Could we please have this behaviour fixed? Maybe this thread should be tagged as [feature-request] instead of [question]? Or should we open a new issue?

**Comment by iainelder** [comment]
I ran into this again and started thinking about how to fix it.

Could we split out the language settings into a separate file, like language.ini or user.ini?

Apart from the language, I can't see a reason to change any of the settings in dbeaver.ini. As far as I can tell it's JVM config. I trust DBeaver to handle that without my intervention.

However, the language is something I do want to change, and I do want my choice to survive upgrades.

If there already is a more appropriate place for user settings, let me know.

It seems that BasePlatformImpl.java is responsible for setting the language. It tries to use the OS language and otherwise uses English. I don't see where it reads dbeaver.ini to get the override!

@serge-rider , any thoughts?

**Comment by SubOptimal** [comment]
As mentioned my @iainelder DBeaver takes the locale from the session it is started or falls back to English in case the locale is not supported.

Depending on your used OS you might create a Shell script in which you set the locale before starting DBeaver.

(adapt the path's and locale to your system)

- e.g. create a shell script `~/bin/dbeaver.sh` with following content
```
#!/bin/bash
export LANG=en_US
~/bin/dbeaver/dbeaver
```

This would start DBeaver with English UI.

Which value you can use for `LANG` depends on the locales configured on your system. Evailable locales can be checked with `locale -a`.

example:
```
~ $ locale -a
C
C.UTF-8
en_GB
en_GB.iso88591
en_GB.utf8
en_US
en_US.iso88591
en_US.utf8
french
fr_FR
fr_FR.iso88591
fr_FR.utf8
POSIX
```

In case none of the `en_*` locales is configured on your system you can try with `C` or `POSIX`.

**Comment by iainelder** [comment]
@SubOptimal, thanks for the workaround. It's easier than remembering how to edit the ini file.

However, I would still like this to be fixed in the program itself.

Can someone tell me which part of the code is responsible for **reading** dbeaver.ini?

In BasePlatformImpl.java I see only code for writing it.

**Comment by SubOptimal** [comment]
@iainelder The way the ini file is determined is defined by the launcher of the Eclipse equinox framework.
https://git.eclipse.org/c/equinox/rt.equinox.framework.git/tree/features/org.eclipse.equinox.executable.feature/library/eclipseConfig.h#n20

The Linux launcher looks for a ini file `launcher_dir/launcher.ini`. (for Windows and MacOS there are some small differences)

One possible way could be to start DBeaver not with the launcher executable but instead with a small wrapper script and pass the `--launcher.ini` option to the launcher [Eclipse runtime options](https://help.eclipse.org/2020-09/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html) and pass the customized ini file location to it.

simplified example of `dbeaver_wrapper.sh` for demonstration purpose
```
#!/bin/bash

if [ -f ~/.local/share/DBeaverData/dbeaver.ini ]; then
   # start DBeaver with the user modified launcher ini file
   $dbeaver_dir/dbeaver --launcher.ini ~/.local/share/DBeaverData/dbeaver.ini
else
   # start DBeaver with the provided launcher ini file
   $dbeaver_dir/dbeaver
fi
```

When the user change the language in DBeaver the file `~/.local/share/DBeaverData/dbeaver.ini` should be amended instead of the `$dbeaver_dir/dbeaver.ini`

**Comment by iainelder** [comment]
> The way the ini file is determined is defined by the launcher of the Eclipse equinox framework.

Ah, wow, I didn't realise that DBeaver depended on Eclipse for this.

DBeaver uses Eclipse's ini file format. It is explained on the [Eclipse wiki](https://wiki.eclipse.org/Eclipse.ini). Basically it lists the command line argumentsused to launch Eclipse.

The `-nl` argument is described in [Eclipse's Runtime arguments documentation](https://help.eclipse.org/2020-09/topic/org.eclipse.platform.doc.isv/reference/misc/runtime-options.html#osginl).

> the name of the locale on which Eclipse platform will run. NL values should follow the standard Java locale naming conventions.

So it turns out that this obscure way of setting the language is inherited from Eclipse. Now I understand why we are discussing scripts that work around the problem: there's no easier way to change this behavior in DBeaver's code.

> One possible way could be to start DBeaver not with the launcher executable but instead with a small wrapper script

That's a neat script, and I can see how it might solve the problem.

What happens if the other arguments change during an upgrade?

Presumably my customized ini file would need to keep track of all the changes to default ini file to avoid errors in the settings for Eclipse startup, launcher library, and JVM.

Sadly that seems no better than the current situation.

**Comment by iainelder** [comment]
As another approach, could the updater be smarter about how it treats the dbeaver.ini file?

Currently the updater current replaces the file content without regard to what was there before.

Would it be possible to parse out the `-nl` line of the old version and include it in the new version in the right place?

**Comment by RamonCar66** [comment]
Hello, I face the same issue. 
In the upgrade instructions it states "Upgrade – don’t be afraid to remove previous DBeaver version – your settings won’t be lost. All settings are kept in the separate folder", 
The fact is that most settings are preserved afte version upgrade, but User Interface language, not.
Maybe a simpler solution could be moving this setting to the "user settings" area ?

Another setting I've to modify after each version upgrade is the java.security policy. I've an old SQL server, and I've to enable an old ecryption method to be able to connect. I understand that this setting is not purely "DBeaver", and maybe more difficult to keep between versions.

Kind regards

**Comment by uslss** [comment]
#12771

**Comment by uslss** [comment]
will be fixed in #12771
